### PR TITLE
spread, tests: add Fedora 29

### PR DESCRIPTION
**Pull request by bboozzoo**
Add Fedora 29 to spread suite and try to workaround different quirks identified in the process.

New issues:
- default umask seems to be `0077` now, causing some issues in snap try tests
- `xdg-desktop-portal` seems to be installed by default
- strace-static (stable is 4.20) does not work with Fedora's kernel (4.19.5 atm), tried the edge version (4.23) but still no luck
- RPM seems to rewrite the shebang line to `#!/usr/bin/sh`, added a workaround for the case when we repack the core snap

The PR includes bits from #6260 

@sergiocazzolato would it be possible to update the image and remove `xdg-desktop-portal` ?

### CODE REVIEW

File: tests/main/degraded/task.yaml [hunk 1]
```diff
@@ -13,7 +13,7 @@ debug: |
         units="$(systemctl --failed --type=$unit_type --no-pager | grep -o -E ".*.$unit_type" | tr '●' ' ' )"
         for unit in $units; do
             echo " -- systemctl status $unit --"
-            systemctl status "$unit"
+            systemctl status "$unit" || true
```
Comment (zyga) [inline tests/main/degraded/task.yaml]: What is the motivation behind this change? When would status fail?

File: tests/main/document-portal-activation/task.yaml [hunk 2]
```diff
@@ -27,6 +27,11 @@ prepare: |
     . "$TESTSLIB"/pkgdb.sh
     #shellcheck source=tests/lib/snaps.sh
     . "$TESTSLIB"/snaps.sh
+    if [[ "$SPREAD_SYSTEM" == fedora-29 ]]; then
```
Comment (zyga) [inline tests/main/document-portal-activation/task.yaml]: It is likely that fedora-30 will have the same property. Perhaps we could have a negative test instead.

File: tests/main/try-with-hooks/task.yaml [hunk 3]
```diff
@@ -1,7 +1,7 @@
 summary: Reproduce a known issue of snap try failure with hooks
 
 prepare: |
-  cp -R "$TESTSLIB"/snaps/basic ./
+  cp -a "$TESTSLIB"/snaps/basic ./
```
Comment (zyga) [inline tests/main/try-with-hooks/task.yaml]: +1 on the change but curious on the motivation that lead to it

File: tests/main/document-portal-activation/task.yaml (hunk 2 above)
Comment (bboozzoo) [reply tests/main/document-portal-activation/task.yaml]: I'd prefer to have the test image updated. Will work with @sergiocazzolato to do that and then I'll drop the `if..` entirely

File: tests/main/try-with-hooks/task.yaml (hunk 3 above)
Comment (bboozzoo) [reply tests/main/try-with-hooks/task.yaml]: Default umask got changed to `0077`. The permissions are already set properly in our tree, but without `-a` they'd get reset to defaults. Then snap verification would bail because the directories are not world readable and so on.

File: tests/main/degraded/task.yaml (hunk 1 above)
Comment (bboozzoo) [reply tests/main/degraded/task.yaml]: It's in the commit message. To recap, `systemctl status <svc>` exits with non-0 status when the service has failed.

File: tests/main/document-portal-activation/task.yaml (hunk 2 above)
Comment (zyga) [reply tests/main/document-portal-activation/task.yaml]: That's unrealistic with comparison to what the regular system contains. I'm mildly opposed to crafting images that pass tests rather than have the test declare what is assumed. I will not block on this but I want to make my position clear.

File: tests/main/degraded/task.yaml (hunk 1 above)
Comment (zyga) [reply tests/main/degraded/task.yaml]: Interesting, thanks.

File: tests/main/document-portal-activation/task.yaml (hunk 2 above)
Comment (bboozzoo) [reply tests/main/document-portal-activation/task.yaml]: I'm looking into this more. A default cloud image does not contain `xdg-desktop-portal` package. This must be pulled in by some of the packages we install in `pkg_dependencies()`.

Comment (bboozzoo) [reply tests/main/document-portal-activation/task.yaml]: An interesting find, apparently when we `dnf install` the built RPM, `dnf` pulls in `flatpak` and `xdg-desktop-portal`. I have no idea how this dependency came to be. Once installed I can actually remove `flatpak` and other packages and `snapd` is not removed. Playing with `dnf deplist` I did not found why exactly this happens. Maybe @Conan-Kudo  can suggest something.

Comment (Conan-Kudo) [reply tests/main/document-portal-activation/task.yaml]: It's most likely a rich weak dependency being triggered. Do we run with weak deps enabled? If so, you could try adding `--setopt=install_weak_deps=False` to not make it happen.

Comment (bboozzoo) [reply tests/main/document-portal-activation/task.yaml]: Thanks, that does the trick

File: tests/lib/pkgdb.sh [hunk 4]
```diff
@@ -172,7 +172,7 @@ distro_install_local_package() {
             apt install $flags "$@"
             ;;
         fedora-*)
-            quiet dnf -y install "$@"
+            quiet dnf -y install --setopt=install_weak_deps=False "$@"
```
Comment (zyga) [inline tests/lib/pkgdb.sh]: I wonder if this will impact anything that only worked because of a weak dependency that was installed before. Let's see.

File: tests/main/auth-errors/task.yaml [hunk 5]
```diff
@@ -5,21 +5,22 @@ systems: [-ubuntu-core-*]
...
     mkdir -p /home/test/.snap
     echo -n "{\"macaroon\":\"yummy\",\"discharges\":[ \"some \"]}" > /home/test/.snap/auth.json
+    chown -R test:test /home/test/.snap
 
 restore: |
     rm -rf /home/test/.snap install.output connect.output
 
 execute: |
     echo "An unauthenticated user cannot install snaps"
-    if su - -c "snap install test-snapd-tools 2>${PWD}/install.output" test; then
```
Comment (zyga) [inline tests/main/auth-errors/task.yaml]: Ouch, how did this ever work before?

File: tests/main/auth-errors/task.yaml [hunk 6]
```diff
@@ -5,21 +5,22 @@ systems: [-ubuntu-core-*]
...
+    if su - -c "snap install test-snapd-tools" test 2>${PWD}/install.output ; then
         echo "Expected error installing snap from unauthenticated account"
         exit 1
     fi
     expected="error: access denied (try with sudo)"
     [ "$(cat install.output)" = "$expected" ]
 
     echo "An unauthenticated user cannot connect plugs to slots"
-    if su - -c "snap connect foo:bar baz:fromp 2>${PWD}/connect.output" test; then
+    if su - -c "snap connect foo:bar baz:fromp" test 2>${PWD}/connect.output ; then
```
Comment (zyga) [inline tests/main/auth-errors/task.yaml]: And this?

File: tests/main/auth-errors/task.yaml (hunk 5 above)
Comment (bboozzoo) [reply tests/main/auth-errors/task.yaml]: Yeah, I was wondering that too.

Comment (zyga) [reply tests/main/auth-errors/task.yaml]: Can you try to understand that? I think there's something fishy in our test suite still.

File: tests/main/document-portal-activation/task.yaml (hunk 2 above)
Comment (Conan-Kudo) [reply tests/main/document-portal-activation/task.yaml]: This makes me wonder if we're not disabling weak deps on all targets for CI...

File: tests/main/auth-errors/task.yaml (hunk 5 above)
Comment (bboozzoo) [reply tests/main/auth-errors/task.yaml]: My bad, I ran su - -c "snap install test-snapd-tools 2>${PWD}/install.output" in spread shell, which ended up creating the file owned by root. Another observation is that although the tests are run by root, /home/gopath is owned by test:test so it all works out in the test user in the end, unless we create files in the test itself and are not careful about changing the permissions. There are more tests that do the su -c ".. ${PWD}>somepath" test pattern, so I'll drop that hunk.

Comment (bboozzoo) [reply tests/main/auth-errors/task.yaml]: Pushed a commit to fix it in other places too. Conveniently, those tests did not execute on Fedora, either by explicit `systems` entry or requiring strict confinement.

File: tests/main/econnreset/task.yaml [hunk 7]
```diff
@@ -38,7 +38,7 @@ execute: |
     iptables -I INPUT -p tcp --match hashlimit --hashlimit-mode srcip,dstip,srcport,dstport --hashlimit-above 512kb/s \
              --hashlimit-name 'econnreset' -j DROP
 
-    su -c "/usr/bin/env SNAPD_DEBUG=1 SNAPD_DEBUG_HTTP=3 snap download --edge test-snapd-huge 2>snap-download.log" test &
```
Comment (stolowski) [inline tests/main/econnreset/task.yaml]: Nothing wrong with this change afacit, but I wonder what was the motivation for it?

Comment (stolowski) [reply tests/main/econnreset/task.yaml]: Ah, ignore me, I see similiar question from zyga above.
